File: Paradygmaty/Lista3/zad2/DHSetup.java
package Lista3.zad2;
import java.security.SecureRandom;
import java.util.ArrayList;
import Lista2.zad2.GaloisField;

class DHSetup<T> {
    private T field;
    private int P;
    private T generator;
    public int getP() {
        return this.P;
    }
    private ArrayList<Integer> findPrimesDividingP(int n) {
        ArrayList<Integer> result = new ArrayList<>();
        boolean marked[] = new boolean[n + 1];
        for(int i = 0; i < marked.length; i++) {
            marked[i] = true;
        }
        for(int i = 2; i <= Math.sqrt(n); i++) {
            if(marked[i]) {
                int j = i * i;
                while(j <= n) {
                    marked[j] = false;
                    j = j + i;
                }
            }
        }
        for(int q = 2; q <= n; q++) {
            if(marked[q]) {
                if(n % q == 0) {
                    result.add(q);
                }
            }
        }
        return result;
    }
    private boolean isGenerator(long a, ArrayList<Integer> primes) {
        for(int prime : primes) {
            long factor = (this.P - 1) / prime;
            long elem = a * factor;
            if(elem == 1) {
                return false;
            }
        }
        return true;
    }
    private long generateRandom(SecureRandom rand, int min, int max) {
        return rand.nextLong(min, max + 1);
    }

    private ArrayList<Integer> decomposeIntoPowersOfTwo(long num) {
        ArrayList<Integer> powers = new ArrayList<>();
        int bitPosition = 0;
        while(num > 0) {
            if((num & 1) == 1) {
                powers.add(1 << bitPosition);
            }
            num >>= 1; // shift bits to right
            bitPosition++; // next bit position
        }
        return powers;
    }
    @SuppressWarnings("unchecked")
    public DHSetup(T field, SecureRandom rand) {
        this.field = field;
        GaloisField myField = null;
        try {
            myField = (GaloisField)field;
        } catch(Exception ex) {
            System.out.println(ex.getMessage());
        }
        this.P = myField.getOrder();
        ArrayList<Integer> primes = this.findPrimesDividingP(this.P - 1);
        T generator = null;
        while(generator == null) {
            long element = this.generateRandom(rand, 1, this.P - 1);
            if(this.isGenerator(element, primes)) {
                System.out.println("Gen is " + element);
                try {
                    generator = (T)new GaloisField(element, this.P);
                } catch(Exception ex) {
                    System.out.println(ex.getMessage());
                }
            }
        }
        this.generator = generator;
    }   

    public T getGenerator() {
        return this.generator;
    }

    @SuppressWarnings("unchecked")
    public T power(T a, long b) {
        GaloisField field = null;
        try {
            field = (GaloisField) a;
        } catch(Exception ex) {
            System.out.println(ex.getMessage());
        }
        long a_val = field.getCharacteristics();
        ArrayList<Integer> powers = this.decomposeIntoPowersOfTwo(b);
        ArrayList<Long> modulus = new ArrayList<>();
        ArrayList<Long> values = new ArrayList<>();
        int j = 0;
        for(int i = 1; i <= powers.get(powers.size() - 1); i *= 2) {
            long modulo;
            if(i == 1) {
                modulo = a_val % this.P;
            } else {
                modulo = (modulus.get(modulus.size() - 1) * modulus.get(modulus.size() - 1)) % this.P;
            }
            modulus.add(modulo);
            if(powers.get(j) == i) {
                j++;
                values.add(modulo);
            }
        }
        long multiply = 1;
        for(long value : values) {
            multiply = (multiply * value) % this.P;
        }
        int value = (int)multiply;
        System.out.println("Val " + value);
        GaloisField gField = null;
        T result = null;
        try {
            gField = new GaloisField(value, this.P);
            result = (T)gField;
        } catch(Exception ex) {
            System.out.println(ex.getMessage());
        }
        return result;
    }
}

File: Paradygmaty/Lista3/zad3/DHSetup.py
from typing import List, TypeVar, Generic, Callable
from numpy import random
import math

T = TypeVar('T')

class DHSetup(Generic[T]):

    def __init__(self, field: T, field_creator: Callable[[int, int], T]):
        self.field = field
        self.P = field.order
        self.field_creator = field_creator
        print(self.P)
        primes = self.findPrimesDividingP(self.P - 1)
        for elem in primes:
            print(f'{elem}')
        gen: T = None
        while gen is None:
            element = self.generateRandom(1, self.P - 1)
            if self.isGenerator(element, primes):
                print(f'Gen is {element}')
                gen = field_creator(element, self.P)
                break
        self.generator = gen

    def findPrimesDividingP(self, n: int) -> list[int]:
        result = []
        marked = []
        for i in range(0, n + 1):
            marked.append(True)
        for i in range(2, int(math.sqrt(n)) + 1):
            if marked[i] == True:
                j = i * i
                while j <= n:
                    marked[j] = False
                    j = j + i
        for j in range(2, n + 1):
            if marked[j] == True: # if j is prime number
                if n % j == 0:
                    result.append(j)
        return result

    def isGenerator(self, a: int, primes) -> bool:
        for prime in primes:
            factor = (self.P - 1) / prime
            elem = a * factor
            if elem == 1:
                return False
        return True

    def generateRandom(self, min: int, max: int):
        return random.randint(min, max + 1)

    def decomposeIntoPowersOfTwo(self, num) -> List[int]:
        powers = []
        bitPosition = 0
        while num > 0:
            if num & 1:
                powers.append(1 << bitPosition)
            num >>= 1
            bitPosition += 1
        return powers
    
    def getGenerator(self) -> T:
        return self.generator
    
    def power(self, a: T, b: int) -> T:
        a_val = a.value
        powers = self.decomposeIntoPowersOfTwo(b)
        modulus = []
        values = []
        j = 0
        i = 1
        while i <= powers[-1]:
            modulo = None
            if i == 1:
                modulo = a_val % self.P
            else:
                modulo = (modulus[-1] * modulus[-1]) % self.P
            modulus.append(modulo)
            if powers[j] == i:
                j += 1
                values.append(modulo) 
            i *= 2
        multiply = 1
        for value in values:
            multiply = (multiply * value) % self.P
        print(f'Wynik {multiply}')
        # return T(multiply, self.P)
        return self.field_creator(multiply, self.P)
    

